thali/thalireplicationmanager.js

```diff
@@ -33,30 +43,83 @@ ThaliReplicationManager.events = {
   START_ERROR: 'startError',
   STOP_ERROR: 'stopError',
   CONNECT_ERROR: 'connectError',
-  DISCONNECT_ERROR: 'disconnectError',
-  SYNC_ERROR: 'syncRetry'
+  DISCONNECT_ERROR: 'disconnectError'
 };
 
 /**
-* Starts the Thali replication manager with the given device name and port number
-* @param {String} deviceName the device name to advertise.
+* Returns the current device's id if available. If not, the provided callback
+* function is saved and called when the id does become available.
+* @param {Function} cb the callback which returns the current device's id.
+*/
+ThaliReplicationManager.prototype.getDeviceIdentity = function (cb) {
+  if(this._deviceIdentityFlag == deviceIdentityFlag.deviceIdentityAvailable) {
+    cb(null, this._deviceName);
+    return;
+  }
+  if(this._deviceIdentityFlag == deviceIdentityFlag.noDeviceIdentitySet) {
+    this._deviceIdentityFlag = deviceIdentityFlag.gettingDeviceIdentity;
+    // save the callback for future
+    this._deviceIdentityListeners.push(cb);
+    cryptomanager.getPublicKeyHash(function (err, publicKeyHash) {
+      if (err) {
+        this._deviceIdentityFlag = deviceIdentityFlag.noDeviceIdentitySet;
+      } else {
+        this._deviceName = publicKeyHash;
+        this._deviceIdentityFlag = deviceIdentityFlag.deviceIdentityAvailable;
+      }
+      informDeviceIdentityListeners.call(this, err, publicKeyHash);
+      this._deviceIdentityListeners = [];
```

Comment:
I think this code has a race condition.

Let's imagine that getPublicKeyHash fails.
So on line 70 we call all the registered CBs.
One of those CBs then immediately calls back into getDeviceIdentity to try again.
Their request will hit line 59 because we already set the flag to noDeviceIdentitySet, they will then have their CB pushed onto _deviceIdentityListeners in line 62 and then they will return when they hit line 63 because it's asynchronous.
At that point we return from line 70 and then call line 71 who immediately erases their enqueued CB.

This is the kind of heiseinbug that drives people insane.

Anytime you write Javascript code you have to always ask yourself "is my function giving up control to anyone else?" because the minute you do you have to think very carefully about re-entrant behavior.

I suspect the least mind bending behavior to fix this is to use something like slice() or equivalent to create a local copy of _deviceIdentityListeners, then set the global version to [] and then run get rid of informDeviceIdentityListeners (since you only use it one place) and just run the foreach that used to be in informDeviceIdentityListeners directly on the local copy. That should make everything work right even with re-entrant behavior. I think.